### ISSUE TITLE
Change teams?

### ISSUE THREAD

**Issue by kevquirk** [kevquirk/512kb.club#386]
Raising a discussion about the team sizes.

Currently its:

* <100kb - Green
* <250kb - Orange
* <512kb - Blue

Wondering if it's worth changing it to:

* <128kb - Green
* <256kb - Orange
* <512kb - Blue

@kmutahar @lrmikestone thoughts?

**Comment by lrmikestone** [comment]
I think it's a valid change. It has minimal impact as there's only 10 sites currently in Orange that would move to Green, and only a single site in Blue that would move to Orange as a result. The new values satisfy my nerd side, but it's worth saying that even though 512kb.club is about sites that maintain a small footprint, it's not necessarily about nerdy things. Non-nerds might find the original values more appropriate, but I don't know if they're represented in numbers high enough for it to matter.

**Comment by kmutahar** [comment]
I really like the idea even if people are non-technical it all divisible by 2 so its really nice

On another note, I have been looking at a way (Still testing) on having just one file and using if conditions to have it sort. what do you think. 

Also as I have just push away for Jekyll to sort the list should we change the file to be in alphabetical order. I belive this would make it easer for future maintenance

**Comment by kevquirk** [comment]
I'm not against moving to one file. Not sure how easy that would be with all three being separate billeted lists, all with a different CSS class for the colours.

If you can work out a way of doing it though, it would certainly be easier.

I think alphabetical works.

Will let you finish the bits you're doing, then we can flip it to the new sizes then. 👍

**Comment by kmutahar** [comment]
Sounds great, the only problem is GTmetric and their limits, if only there was a CLI that can provide the size

**Comment by kevquirk** [comment]
Yeah, I'm not sure if one of the other tools provides an API to do this kinda thing. I imagine there would be a charge for it though.

Doing a valid response check would also be cool, so we can remove and sites that are down. Once to explore, I think.

**Comment by kevquirk** [comment]
This could work for the check...

https://www.darklaunch.com/count-size-of-page-in-bytes-using-curl.html

**Comment by mrxkon** [comment]
Sorry for bumping in here but it's an interesting discussion as I saw your curl idea ( and as I was also confused of which uncompressed value you where after when I submitted my site recently ).

I was under the impression that we're after the overall size of a page (including assets etc) not just the `/`.

As an example the curl you mentioned would work correctly:
This would return 22.904kb .
`curl --silent "https://xkon.dev" | wc --bytes | awk '{ byte =$1 /1000; print byte " kb" }'`

But it only takes under account the returned html essentially not "everything" that the page would load.

According to GTMetrix (setting as an example my page as well) while `/` is actually ~22.4kb uncompressed the overall page (due to some images) is 60.8kb, which was the value that I submitted.

So I'm confused now of what the Club  is after in reality and if it needs to be more clear on the instruction of submitting a site. 

The instructions refer to "Get the UNCOMPRESSED size of your website's homepage", homepage can mean with every asset included as well or just the / which is a totally different number :) .

**Comment by kevquirk** [comment]
@mrxkon please don't be sorry, the conversation is good. What we ask for is the complete webpage size, so all assets. I hadn't tested the curl command yet, so wasn't sure if it would work. Clearly it wouldn't.  So thanks for testing that for us. 😊

I think we do need a method of regularly testing if a site is a) up and b) the size. I just don't know how best to do that yet.

**Comment by kmutahar** [comment]
I have a question about the current size and the teams for #390 

```
Green
>= 0 and <= 100
Orange
> 100 and  <= 250
Blue
> 250 and <= 512
```

or should it be 

```
Green
>= 0 and < 100
Orange
>= 100 and  < 250
Blue
>= 250 and <= 512
```

it's the sites that are on the border where should they be located?

**Comment by kevquirk** [comment]
I think it should be (and this is just my thought) is it's **exactly** 100kb, then it's green team. If it's 100.1, then it's orange.

**Comment by Lex-2008** [comment]
If you let me chime in, as a person whose website is _just a bit_ under 100kb and is planning only to grow (since it's a one long list of all my 132 blog posts), I'm all in favour of increasing the limit to get into the green team :)

On a serious note, regarding periodically checking sizes of all sites: have you considered using webdriver?

In my testing I noticed that numbers shown on gtmetrix.com (total size of all requests) matches what I see in Chrome DevTools (and, indeed, they say they use Chrome).

It looks like it might be possible to drive Chrome using WebDriver and monitor data received over the network:
https://chromedevtools.github.io/devtools-protocol/1-2/Network/#event-dataReceived

Does anyone see why it might be a bad idea to have a local instance of Chrome slowly crawling list of all websites, recording their sizes? If none else wants, I can try to make such script in my spare time.

**Comment by kmutahar** [comment]
@Lex-2008 we haven't found anything that's accurate like GTmetrix. I will check out Webdrive. 

If you do have an idea how it can be implemented let us know. It would be nice to have a script that one can use every once in a while to test the sites.

**Comment by kevquirk** [comment]
+1 to @kmutahar comment above. A script would be really nice!

**Comment by Lex-2008** [comment]
Based on a [gist][] by @rengler33, here's a script which does that:

[page_size.py.gz](https://github.com/kevquirk/512kb.club/files/6879314/page_size.py.gz)

It requires Chrome, [chromedriver][] in your path, and python selenium installed either via [pip][] or your OS [package manager][arch].

You can start it from command-line an pass it URLs as arguments, like this:

````
$ python page_size.py https://512kb.club/ http://alexey.shpakovsky.ru https://www.yeetpc.com/
````

It will open the browser, load each page into it, and print back size of each page, like this:

````
$ python page_size.py https://512kb.club/ http://alexey.shpakovsky.ru https://www.yeetpc.com/
https://512kb.club/: 69736 bytes = 68.1 kb
http://alexey.shpakovsky.ru: 90267 bytes = 88.2 kb
https://www.yeetpc.com/: 530334 bytes = 517.9 kb
````

[gist]: https://gist.github.com/rengler33/f8b9d3f26a518c08a414f6f86109863c
[chromedriver]: https://chromedriver.chromium.org/
[pip]: https://pypi.org/project/selenium/
[arch]: https://archlinux.org/packages/community/x86_64/python-selenium/

What do you think about it? Numbers look pretty close to those reported by GTmetrix. I'm planning to expand it to read list of sites from the _data/sites.yml file and save new numbers into it.

**Comment by Lex-2008** [comment]
Updated version of the script above:

[script.py.gz](https://github.com/kevquirk/512kb.club/files/6932348/script.py.gz)

To update sites.yml file while keeping its format it now also uses "ruamel.yaml" Python library available via [pip] or a [package manager][arch] of choice.

[pip]: https://pypi.org/project/ruamel.yaml/
[arch]: https://archlinux.org/packages/community/any/python-ruamel-yaml/

It reads a range of sites from sites.yml file, loads each of them updating the sites.yml file, printing old and new size, and "notable changes" section at the end when a page size changed by 10% or more, site "team" changed, or site grew too big to be included to 512kb.club, like this:

````
_data$ time python script.py sites.yml 60-70
cyberarm.dev: 10.7kb to 11.4kb
cycloneblaze.net: 41.9kb to 43.9kb
dafne.cc: 16.5kb to 17.8kb
danielcuttridge.com: 26.1kb to 20.5kb
dariusz.wieckiewicz.org: 257.5kb to 502.1kb
davidrutland.com: 164.0kb to 167.3kb
decentnet.github.io: 31.2kb
degruchy.org: 3.2kb to 1147.1kb
dejalavidavolar.org: 96.1kb to 119.4kb
dev-nonsense.com: 5.5kb


*** NOTABLE CHANGES ***:

danielcuttridge.com: big size change: from 26.1kb to 20.5kb, recheck: <https://gtmetrix.com/?url=https://danielcuttridge.com/>
dariusz.wieckiewicz.org: big size change: from 257.5kb to 502.1kb, recheck: <https://gtmetrix.com/?url=https://dariusz.wieckiewicz.org/>
degruchy.org: site is too big for 512kb.club: 1147.1kb, recheck: <https://gtmetrix.com/?url=https://degruchy.org/>
dejalavidavolar.org: team changed: from green to orange (119.4kb), recheck: <https://gtmetrix.com/?url=https://dejalavidavolar.org/>

real    0m49.205s
user    0m11.359s
sys     0m3.661s
````

As you see, on average it takes about 5 seconds to get a size of one site, so about half an hour to process all 300 links in sites.yml file (not counting time of manually checking results and contacting site owners)

I'm trying to make numbers produced by this script to be as close as possible to what GTmetrix shows, and so far faced two issues:
*  blog.larsgerber.ch uses "Roboto Mono" font, which is installed on GTmetrix machines (so they don't download it), but not on mine; and even after I install it, Chrome says it's available but still downloads it anew, thus inflating pagesize number.
*  canistop.net has a "link rel="preload" tag for metadata.json, and later does "fetch" this file with javascript. And for some reason, on my machine Chrome downloads this file twice, while on GTmetrix they download it only once - thus, again, different pagesize number.

Any thoughts, opinions about this? I think we can either just exclude such problematic sites from this script, or give them some "discount" when size is calculated by this script, in order to match the GTmetrix value.

Also, it looks like GTmetrix has a free API for 16 sites a day, so it's possible to make a different script which will update sites.yml file in batches of 16 sites a day :)

What do you think?
